PostHog/posthog
# fix(insights): Cancel running queries if user aborts

## Problem

It's very easy for a user to fire off a bunch of queries quickly. See [this metabase query (internal)](https://metabase.posthog.net/question#eyJkYXRhc2V0X3F1ZXJ5Ijp7InR5cGUiOiJuYXRpdmUiLCJuYXRpdmUiOnsicXVlcnkiOiJzZWxlY3QgbWVtLyh4LzEwMDAwMDAwMDApKjEwMCwgKiBmcm9tIChcbnNlbGVjdCBjb3VudCgxKSBhcyBjLCBzdW0obWVtb3J5X3VzYWdlKS8xMDAwMDAwMDAwIGFzIG1lbSwgc3VtKHN1bShtZW1vcnlfdXNhZ2UpKSBvdmVyICgpIGFzIHgsIGF2ZyhxdWVyeV9kdXJhdGlvbl9tcykvMTAwMCBhcyBkdXJhdGlvbiwgYXJyYXlDb21wYWN0KGdyb3VwQXJyYXkodHlwZSkpIGFzIHN0YXR1cywgYXJyYXlDb21wYWN0KGdyb3VwQXJyYXkocXVlcnlfc3RhcnRfdGltZSkpLCBxdWVyeVxuZnJvbSBjbHVzdGVyQWxsUmVwbGljYXMocG9zdGhvZywgc3lzdGVtLCBxdWVyeV9sb2cpIHdoZXJlXG4gICAgcXVlcnlfc3RhcnRfdGltZSA-ICcyMDIyLTA4LTI0VDE2OjAwOjAwJyBhbmQgcXVlcnlfc3RhcnRfdGltZSA8ICcyMDIyLTA4LTI0VDE2OjA1OjAwJyBhbmQgdHlwZSA-IDEgYW5kIHF1ZXJ5IG5vdCBsaWtlICdJTlNFUlQgSU5UTyUnXG5ncm91cCBieSBxdWVyeSBvcmRlciBieSBtZW0gZGVzY1xuKSIsInRlbXBsYXRlLXRhZ3MiOnt9fSwiZGF0YWJhc2UiOjJ9LCJkaXNwbGF5IjoidGFibGUiLCJ2aXN1YWxpemF0aW9uX3NldHRpbmdzIjp7fX0=) for an example where within a 5 minute window a user fired off a bunch of very similar queries very quickly, which caused those queries to fail.

## Changes

- Send KILL QUERY command if the user aborts the query (by changing some of the parameters, or navigating way)

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

- set `cancel-running-queries` feature flag
- Add `sleep(3)` as last field in `AGGREGATE_SQL` variable in posthog/queries/trends/sql.py
- Quickly duplicate graph series
- See /cancel being fired
